Podcast Re-subscription fixes

bjtitus opened the pull request:
Fixes [PCIOS-175](https://linear.app/a8c/issue/PCIOS-175/ios-unsubscribed-podcasts-reappear-after-update)

There are two primary fix attempts here:

* Make sure we are fetching and passing `shouldSubscribe` when we call `serverPodcastManager.addFromUuid`. Otherwise, we could potentially subscribe to a Podcast which the server already told us is unsubscribed.
* When importing a podcast, we seemed to be setting `subscribed` based on the `isDeleted` value. It seems like we should also use `subscribed` if only it exists. I'm not 100% sure on why we ever use `isDeleted`, but I decided to keep that for now. It only happens in specific cases where `checkIsDeleted` is `true` but this might be something to look at in the future.

So both fixes really come down to using the server's `subscribed` value where we weren't before.

## To test

* Use two devices to test syncing
* Subscribe to a podcast on one device
* Refresh in the Profile tab to sync
* Unsubscribe on another device
* Refresh in the Profile tab to sync
* Verify that podcasts are properly synced and removed from lists


## Checklist

- [x] I have considered if this change warrants user-facing release notes and have added them to `CHANGELOG.md` if necessary.
- [x] I have considered adding unit tests for my changes.
- [x] I have updated (or requested that someone edit) [the spreadsheet](https://docs.google.com/spreadsheets/d/107jqrutZhU0fVZJ19SBqxxVKbV2NWSdQC9MFYdLiAxc/edit?usp=sharing) to reflect any new or changed analytics.

bjtitus committed:
Add unit tests for podcast import

bjtitus committed:
Ensure sync progress notifications always fire

Ensure sync progress notifications always fire: Wraps importPodcast’s progress counter updates in a defer so every exit path still posts syncProgressPodcastUpto and syncProgressPodcastCount, preserving UI accuracy when we return early.

bjtitus committed:
Refactor deleted podcast handling

This allows us to more easily bail out with early returns in cases where we may want to (podcast is deleted or unsubscribed)

bjtitus committed:
Respect server subscription intent for new podcasts

Respect server subscription intent for new podcasts: Adds a derived shouldSubscribe flag and passes it through to addFromUuid, preventing re-subscribing podcasts the server reports as unsubscribed.

We could also short-circuit the addFromUUID call on shouldSubscribe being false but I figured we may be better off keeping that functionality as is.

bjtitus committed:
Honor server subscription flag when updating existing podcasts

Uses the subscribed value instead of updating to isDeleted. This should help in case we only receive subscribed and not isDeleted. We won't want to set subscribed to 0 when isDeleted is missing but subscribed is 0

dangermattic commented:
<!--
  0 Errors
  0 Warnings
  1 Message: This PR is still a Draft: some...
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">This PR is still a Draft: some checks will be skipped.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

bjtitus committed:
Update test to test for added unsubscribed podcast

This behavior was changed a bit last minute to keep the podcast in the DB with the subscribed value set rather than skipping adding it.

pocketcasts commented:
<!-- REUSE_ID: 3675df1f-c821-42fe-b581-87f3ea789cdb -->

Version `7.100` has now entered code-freeze, so the milestone of this PR has been updated to `7.101`.

SergioEstevao commented:
@bjtitus and I tested this together and we think this PR needs some extra checks to see if all the logic is solid taking in account our server implementation for sync

pocketcasts commented:
<!-- REUSE_ID: c583978d-47e3-47a6-9e66-e7df0d55f76c -->

Version `7.101` has now entered code-freeze, so the milestone of this PR has been updated to `7.102`.

SergioEstevao commented:
@bjtitus are you still looking into this? Should I remove myself as reviewer?

pocketcasts commented:
<!-- REUSE_ID: d8cef232-8a26-41d7-b6ec-0ddc8c9bf439 -->

Version `8.0` has now entered code-freeze, so the milestone of this PR has been updated to `8.1`.

pocketcasts commented:
<!-- REUSE_ID: e9105e32-064e-4b33-966a-e8e1a0f0c527 -->

Version `8.1` has now entered code-freeze, so the milestone of this PR has been updated to `8.2`.

pocketcasts commented:
<!-- REUSE_ID: d1ee8fbf-17d8-46cf-bdc8-6511ebbc8c2a -->

Version `8.2` has now entered code-freeze, so the milestone of this PR has been updated to `8.3`.

pocketcasts commented:
<!-- REUSE_ID: 02d30ed9-2c38-44d7-ae9c-baa933b102ca -->

Version `8.3` has now entered code-freeze, so the milestone of this PR has been updated to `8.4`.

pocketcasts commented:
<!-- REUSE_ID: 94390a24-76ed-4741-9966-1de12c70edda -->

Version `8.4` has now entered code-freeze, so the milestone of this PR has been updated to `8.5`.

pocketcasts commented:
<!-- REUSE_ID: 02845eaa-079f-4404-b234-bbd0b48dae43 -->

Version `8.5` has now entered code-freeze, so the milestone of this PR has been updated to `8.6`.

pocketcasts commented:
<!-- REUSE_ID: 6e7ed31a-4a91-4ead-96f7-c14c5b3a97ff -->

Version `8.6` has now entered code-freeze, so the milestone of this PR has been updated to `8.7`.

pocketcasts commented:
<!-- REUSE_ID: 8725a6d3-013d-4037-89dc-4e620acc33ce -->

Version `8.7` has now entered code-freeze, so the milestone of this PR has been updated to `8.8`.

pocketcasts commented:
<!-- REUSE_ID: 65edbe3f-a5f0-4c45-9ab8-e1c9cbbda28c -->

Version `8.8` has now entered code-freeze, so the milestone of this PR has been updated to `8.9`.

pocketcasts commented:
<!-- REUSE_ID: d360311f-806b-440d-b516-517b6023dbe2 -->

Version `8.9` has now entered code-freeze, so the milestone of this PR has been updated to `8.10`.

pocketcasts commented:
<!-- REUSE_ID: 885b2769-c3d2-4f75-8cbe-0f301b0126a1 -->

Version `8.10` has now entered code-freeze, so the milestone of this PR has been updated to `8.11`.

pocketcasts commented:
<!-- REUSE_ID: f4fadced-8e25-47d9-bdd4-02f69a8cad40 -->

Version `8.11` has now entered code-freeze, so the milestone of this PR has been updated to `8.12`.

pocketcasts commented:
<!-- REUSE_ID: 226a3b9c-21ed-44e2-8a76-34b1d03ecc7a -->

Version `8.12` has now entered code-freeze, so the milestone of this PR has been updated to `8.13`.

pocketcasts commented:
<!-- REUSE_ID: 9456de40-fe04-4f33-8418-004b9afd72a9 -->

Version `8.13` has now entered code-freeze, so the milestone of this PR has been updated to `8.14`.

pocketcasts commented:
<!-- REUSE_ID: a68b6ca8-306c-4712-8e01-684fe96ba688 -->

Version `8.14` has now entered code-freeze, so the milestone of this PR has been updated to `8.15`.
